src/rules/01_cutadapt.snake

```diff
@@ -23,18 +24,33 @@ rule extract_umi:
         err = RESULTS + "/logs/{sample}_extract_umi.err"
     threads:
         1
-    shell:
-        """
-        umi_tools extract \
-            -I '{input[1]}' \
-            --extract-method=string \
-            --bc-pattern={params.umi_pat} \
-            --read2-in='{input[0]}' \
-            --read2-out='{output.R1}' \
-            -L '{output.stats}' \
-            | gzip \
-            > '{output.R2}'
-        """
+    run:
```

Comment:
Code included with the `run:` directive will not get run within the Docker container. As a workaround I wrote all my python code as functions that are saved in `rules.py` and then call these functions by passing an `EOF` block to the `python` command.

Here is an example: https://github.com/sheridar/find-pauses/blob/9a02a9e31638a403650aa11bcad94840a30d6b20/src/rules/05_gene_subsample.snake#L31